src/DotNetty.Codecs/Base64/Base64.cs

```diff
@@ -0,0 +1,205 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace DotNetty.Codecs.Base64
+{
+    using System;
+    using DotNetty.Buffers;
+
+    public class Base64
+    {
+        const int MAX_LINE_LENGTH = 76;
+        const byte EQUALS_SIGN = (byte)'='; //pad
+        const byte NEW_LINE = (byte)'\n';
+        const sbyte WHITE_SPACE_ENC = -5; // Indicates white space in encoding
+        const sbyte EQUALS_SIGN_ENC = -1; // Indicates equals sign in encoding
+
+        public static IByteBuffer Encode(IByteBuffer src)
+        {
+            return Encode(src, Base64Dialect.STANDARD);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, Base64Dialect dialect)
+        {
+            return Encode(src, src.ReaderIndex, src.ReadableBytes, dialect.breakLinesByDefault, dialect);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, bool breakLines, Base64Dialect dialect)
+        {
+            return Encode(src, src.ReaderIndex, src.ReadableBytes, breakLines, dialect);
+        }
+
+        public static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect)
+        {
+            return Encode(src, offset, length, breakLines, dialect, src.Allocator);
+        }
+
+        public unsafe static IByteBuffer Encode(IByteBuffer src, int offset, int length, bool breakLines, Base64Dialect dialect, IByteBufferAllocator alloctor)
+        {
+            if (src == null)
+            {
+                throw new NullReferenceException("src is null");
+            }
+            if (dialect.alphabet == null)
+            {
+                throw new NullReferenceException("alphabet is null");
+            }
+            if (dialect.alphabet.Length != 64)
+            {
+                throw new Exception(String.Format("alphabet.Length:{0}, not 64!", dialect.alphabet.Length));
+            }
+            if (length <= 0) return Unpooled.Empty;
+
+            int outLength = length / 3 * 4;
+            outLength += length % 3 > 0 ? 4 : 0;
+            outLength += breakLines ? outLength / MAX_LINE_LENGTH : 0;
+            IByteBuffer dest = alloctor.Buffer(outLength).WithOrder(src.Order);
+            //avoid unnecessary range checking
+            fixed (byte* alphabet = dialect.alphabet, srcArray = src.Array, destArray = dest.Array)
```

Comment:
>Array [](start = 69, length = 5)

IByteBuffer.Array may throw if buffer is backed by more than one array. If you want to have a direct access to data in memory that backs the byte buffer, the best approach is to use a combination of `IByteBuffer.IoBufferCount` and then:
- either `IByteBuffer.GetIoBuffer(int, int) if the count is 1.
- or `IByteBuffer.GetIoBuffers(int, int)` if count is > 1.
- or use `IByteBuffer` methods for data management (e.g. GetByte, SetByte) if count is -1 (which means that the buffer's underlying structure cannot be represented as an array).
The "use IByteBuffer methods for data management" approach must be there as a common fallback in any case.